File: android/app/src/main/java/com/example/ubapp/games/mafia/MafiaEngine.kt
package com.example.ubapp.games.mafia

import kotlin.random.Random

enum class MafiaPhase { LOBBY, NIGHT, DAY_REVEAL, DAY_VOTE, GAME_OVER }
enum class MafiaWinner { MAFIA, TOWN }

enum class MafiaRole(val displayName: String, val tagline: String) {
    MAFIA("Mafia", "Eliminate the town. Coordinate with your fellow mafia at night."),
    DOCTOR("Doctor", "Save one player each night. You can self-save once per game."),
    VILLAGER("Villager", "You have no special ability. Use your vote during the day.");

    val isTown: Boolean get() = this != MAFIA
    val hasNightAction: Boolean get() = this == MAFIA || this == DOCTOR
}

class MafiaPlayer(val id: String, val name: String, val isHost: Boolean) {
    var role: MafiaRole? = null
    var alive: Boolean = true
}

data class MafiaNightOutcome(val killedId: String?, val savedId: String?)
data class MafiaDayOutcome(val eliminatedId: String?, val tally: Map<String, Int>)

/**
 * Pure game logic. The server adapter is responsible for collecting messages
 * and feeding them in; the engine never touches network code.
 */
class MafiaEngine(private val rng: Random = Random.Default) {
    val tutorialVote = com.example.ubapp.tutorials.TutorialVote()
    val players: MutableMap<String, MafiaPlayer> = linkedMapOf()
    var phase: MafiaPhase = MafiaPhase.LOBBY
    var day: Int = 0
    var winner: MafiaWinner? = null

    private val mafiaVotes = mutableMapOf<String, String>()
    private var doctorTarget: String? = null
    private var doctorSelfSaveUsed = false

    val dayVotes: MutableMap<String, String?> = mutableMapOf()
    var lastNight: MafiaNightOutcome? = null
    var lastDay: MafiaDayOutcome? = null

    fun addPlayer(id: String, name: String, isHost: Boolean = false): MafiaPlayer {
        val p = MafiaPlayer(id, name, isHost)
        players[id] = p
        return p
    }
    fun removePlayer(id: String) { if (phase == MafiaPhase.LOBBY) players.remove(id) }

    val canStart: Boolean get() = phase == MafiaPhase.LOBBY && players.size >= 4

    fun start() {
        if (!canStart) return
        val ids = players.keys.toMutableList().also { it.shuffle(rng) }
        val mafiaCount = (ids.size / 4).coerceIn(1, ids.size - 2)
        for (i in ids.indices) {
            val p = players[ids[i]]!!
            p.role = when {
                i < mafiaCount -> MafiaRole.MAFIA
                i == mafiaCount -> MafiaRole.DOCTOR
                else -> MafiaRole.VILLAGER
            }
        }
        phase = MafiaPhase.NIGHT
        day = 1
    }

    val aliveMafia get() = players.values.filter { it.alive && it.role == MafiaRole.MAFIA }
    val aliveDoctors get() = players.values.filter { it.alive && it.role == MafiaRole.DOCTOR }
    val alive get() = players.values.filter { it.alive }
    val dead get() = players.values.filter { !it.alive }

    fun submitMafiaVote(voterId: String, targetId: String): Boolean {
        if (phase != MafiaPhase.NIGHT) return false
        val voter = players[voterId] ?: return false
        if (!voter.alive || voter.role != MafiaRole.MAFIA) return false
        val target = players[targetId] ?: return false
        if (!target.alive) return false
        mafiaVotes[voterId] = targetId
        return isNightReady()
    }

    fun submitDoctorTarget(doctorId: String, targetId: String): Boolean {
        if (phase != MafiaPhase.NIGHT) return false
        val doc = players[doctorId] ?: return false
        if (!doc.alive || doc.role != MafiaRole.DOCTOR) return false
        val target = players[targetId] ?: return false
        if (!target.alive) return false
        if (targetId == doctorId && doctorSelfSaveUsed) return false
        doctorTarget = targetId
        return isNightReady()
    }

    private fun isNightReady(): Boolean {
        val mafiaSubmitted = aliveMafia.all { mafiaVotes.containsKey(it.id) }
        val doctorSubmitted = aliveDoctors.isEmpty() || doctorTarget != null
        return mafiaSubmitted && doctorSubmitted
    }

    fun resolveNight(): MafiaNightOutcome {
        val tally = mutableMapOf<String, Int>()
        for (t in mafiaVotes.values) tally[t] = (tally[t] ?: 0) + 1
        var killTarget = uniqueMax(tally)
        var saved: String? = null
        val dt = doctorTarget
        if (dt != null && dt == killTarget) {
            saved = dt
            if (dt == aliveDoctors.firstOrNull()?.id) doctorSelfSaveUsed = true
            killTarget = null
        }
        killTarget?.let { players[it]?.alive = false }
        val out = MafiaNightOutcome(killedId = killTarget, savedId = saved)
        lastNight = out
        mafiaVotes.clear(); doctorTarget = null
        phase = MafiaPhase.DAY_REVEAL
        return out
    }

    fun advanceToDayVote() {
        if (phase != MafiaPhase.DAY_REVEAL) return
        dayVotes.clear()
        if (checkWin()) return
        phase = MafiaPhase.DAY_VOTE
    }

    fun submitDayVote(voterId: String, targetId: String?): Boolean {
        if (phase != MafiaPhase.DAY_VOTE) return false
        val voter = players[voterId] ?: return false
        if (!voter.alive) return false
        if (targetId != null) {
            val t = players[targetId] ?: return false
            if (!t.alive) return false
        }
        dayVotes[voterId] = targetId
        return alive.all { dayVotes.containsKey(it.id) }
    }

    fun resolveDay(): MafiaDayOutcome {
        val tally = mutableMapOf<String, Int>()
        for (t in dayVotes.values) if (t != null) tally[t] = (tally[t] ?: 0) + 1
        val candidate = uniqueMax(tally)
        var eliminated: String? = null
        if (candidate != null) {
            val max = tally[candidate]!!
            if (max * 2 > alive.size) eliminated = candidate
        }
        eliminated?.let { players[it]?.alive = false }
        val out = MafiaDayOutcome(eliminatedId = eliminated, tally = tally)
        lastDay = out
        if (checkWin()) return out
        day += 1
        phase = MafiaPhase.NIGHT
        return out
    }

    private fun checkWin(): Boolean {
        val liveMafiaCount = aliveMafia.size
        val liveTownCount = alive.count { it.role != MafiaRole.MAFIA }
        if (liveMafiaCount == 0) { winner = MafiaWinner.TOWN; phase = MafiaPhase.GAME_OVER; return true }
        if (liveMafiaCount >= liveTownCount) { winner = MafiaWinner.MAFIA; phase = MafiaPhase.GAME_OVER; return true }
        return false
    }

    private fun uniqueMax(tally: Map<String, Int>): String? {
        var max = 0
        val tied = mutableListOf<String>()
        for ((id, c) in tally) {
            if (c > max) { max = c; tied.clear(); tied.add(id) }
            else if (c == max) tied.add(id)
        }
        return if (tied.size == 1) tied[0] else null
    }
}

File: ios/Ubapp/Games/Mafia/MafiaEngine.swift
import Foundation

enum MafiaPhase { case lobby, night, dayReveal, dayVote, gameOver }
enum MafiaWinner { case mafia, town }

enum MafiaRole: String {
    case mafia, doctor, villager
    var displayName: String {
        switch self { case .mafia: "Mafia"; case .doctor: "Doctor"; case .villager: "Villager" }
    }
    var tagline: String {
        switch self {
        case .mafia: "Eliminate the town. Coordinate with your fellow mafia at night."
        case .doctor: "Save one player each night. You can self-save once per game."
        case .villager: "You have no special ability. Use your vote during the day."
        }
    }
    var isTown: Bool { self != .mafia }
    var hasNightAction: Bool { self == .mafia || self == .doctor }
}

final class MafiaPlayer {
    let id: String
    let name: String
    let isHost: Bool
    var role: MafiaRole?
    var alive: Bool = true
    init(id: String, name: String, isHost: Bool) {
        self.id = id; self.name = name; self.isHost = isHost
    }
}

struct NightOutcome { let killedId: String?; let savedId: String? }
struct DayOutcome { let eliminatedId: String?; let tally: [String: Int] }

/// Pure game logic. The server adapter is responsible for collecting
/// messages and feeding them in; the engine never touches network code.
final class MafiaEngine {
    private var rng: any RandomNumberGenerator

    private(set) var players: [String: MafiaPlayer] = [:]
    var phase: MafiaPhase = .lobby
    var day: Int = 0
    var winner: MafiaWinner?

    private var mafiaVotes: [String: String] = [:]
    private var doctorTarget: String?
    private var doctorSelfSaveUsed = false

    var dayVotes: [String: String?] = [:]
    var lastNight: NightOutcome?
    var lastDay: DayOutcome?

    init(rng: any RandomNumberGenerator = SystemRandomNumberGenerator()) {
        self.rng = rng
    }

    // MARK: Lobby
    @discardableResult
    func addPlayer(id: String, name: String, isHost: Bool = false) -> MafiaPlayer {
        let p = MafiaPlayer(id: id, name: name, isHost: isHost)
        players[id] = p
        return p
    }

    func removePlayer(_ id: String) {
        guard phase == .lobby else { return }
        players[id] = nil
    }

    var canStart: Bool { phase == .lobby && players.count >= 4 }

    func start() {
        guard canStart else { return }
        var ids = Array(players.keys).shuffled(using: &rng)
        let mafiaCount = max(1, min(ids.count - 2, ids.count / 4))
        for i in 0..<ids.count {
            let p = players[ids[i]]!
            if i < mafiaCount { p.role = .mafia }
            else if i == mafiaCount { p.role = .doctor }
            else { p.role = .villager }
        }
        phase = .night
        day = 1
    }

    // MARK: Night
    var aliveMafia: [MafiaPlayer] { players.values.filter { $0.alive && $0.role == .mafia } }
    var aliveDoctors: [MafiaPlayer] { players.values.filter { $0.alive && $0.role == .doctor } }
    var alive: [MafiaPlayer] { players.values.filter { $0.alive } }
    var dead: [MafiaPlayer] { players.values.filter { !$0.alive } }

    @discardableResult
    func submitMafiaVote(voterId: String, targetId: String) -> Bool {
        guard phase == .night,
              let voter = players[voterId], voter.alive, voter.role == .mafia,
              let target = players[targetId], target.alive else { return false }
        mafiaVotes[voterId] = targetId
        return isNightReady()
    }

    @discardableResult
    func submitDoctorTarget(doctorId: String, targetId: String) -> Bool {
        guard phase == .night,
              let doc = players[doctorId], doc.alive, doc.role == .doctor,
              let target = players[targetId], target.alive else { return false }
        if targetId == doctorId && doctorSelfSaveUsed { return false }
        doctorTarget = targetId
        return isNightReady()
    }

    private func isNightReady() -> Bool {
        let mafiaSubmitted = aliveMafia.allSatisfy { mafiaVotes[$0.id] != nil }
        let doctorSubmitted = aliveDoctors.isEmpty || doctorTarget != nil
        return mafiaSubmitted && doctorSubmitted
    }

    @discardableResult
    func resolveNight() -> NightOutcome {
        var tally: [String: Int] = [:]
        for t in mafiaVotes.values { tally[t, default: 0] += 1 }
        let killTarget = uniqueMax(tally)

        var saved: String?
        var finalKill = killTarget
        if let dt = doctorTarget, dt == killTarget {
            saved = dt
            if dt == aliveDoctors.first?.id { doctorSelfSaveUsed = true }
            finalKill = nil
        }
        if let k = finalKill { players[k]?.alive = false }

        let out = NightOutcome(killedId: finalKill, savedId: saved)
        lastNight = out
        mafiaVotes.removeAll()
        doctorTarget = nil
        phase = .dayReveal
        return out
    }

    func advanceToDayVote() {
        guard phase == .dayReveal else { return }
        dayVotes.removeAll()
        if checkWin() { return }
        phase = .dayVote
    }

    // MARK: Day
    @discardableResult
    func submitDayVote(voterId: String, targetId: String?) -> Bool {
        guard phase == .dayVote,
              let voter = players[voterId], voter.alive else { return false }
        if let t = targetId, let p = players[t], !p.alive { return false }
        if let t = targetId, players[t] == nil { return false }
        dayVotes[voterId] = targetId
        return alive.allSatisfy { dayVotes[$0.id] != nil }
    }

    @discardableResult
    func resolveDay() -> DayOutcome {
        var tally: [String: Int] = [:]
        for case let .some(t) in dayVotes.values { tally[t, default: 0] += 1 }
        let candidate = uniqueMax(tally)
        var eliminated: String?
        if let c = candidate, let maxCount = tally[c], maxCount * 2 > alive.count {
            eliminated = c
        }
        if let e = eliminated { players[e]?.alive = false }
        let out = DayOutcome(eliminatedId: eliminated, tally: tally)
        lastDay = out
        if checkWin() { return out }
        day += 1
        phase = .night
        return out
    }

    private func checkWin() -> Bool {
        let liveMafiaCount = aliveMafia.count
        let liveTownCount = alive.filter { $0.role != .mafia }.count
        if liveMafiaCount == 0 { winner = .town; phase = .gameOver; return true }
        if liveMafiaCount >= liveTownCount { winner = .mafia; phase = .gameOver; return true }
        return false
    }

    private func uniqueMax(_ tally: [String: Int]) -> String? {
        var maxCount = 0
        var tied: [String] = []
        for (id, c) in tally {
            if c > maxCount { maxCount = c; tied = [id] }
            else if c == maxCount { tied.append(id) }
        }
        return tied.count == 1 ? tied[0] : nil
    }
}
